pkg/controller/replica.go

```diff
@@ -0,0 +1,338 @@
+	}
+
+	// get all replica attachments for the given volume
+	volReq, err := createLabelRequirements(azureutils.VolumeNameLabel, volumeName)
+	if err != nil {
+		return err
+	}
+	roleReq, err := createLabelRequirements(azureutils.RoleLabel, string(v1alpha1.ReplicaRole))
+	if err != nil {
+		return err
+	}
+	labelSelector := labels.NewSelector().Add(*volReq, *roleReq)
+	azVolumeAttachments := &v1alpha1.AzVolumeAttachmentList{}
+	err = r.client.List(ctx, azVolumeAttachments, &client.ListOptions{LabelSelector: labelSelector})
+	if err != nil {
+		klog.Errorf("failed to list AzVolumeAttachment: %v", err)
+		return err
+	}
+
+	desiredReplicaCount, currentReplicaCount := azVolume.Spec.MaxMountReplicaCount, len(azVolumeAttachments.Items)
+	klog.Infof("control number of replicas for volume (%s): desired=%d,\tcurrent:%d", azVolume.Spec.UnderlyingVolume, desiredReplicaCount, currentReplicaCount)
+
+	// if the azVolume is marked deleted, do no create more azvolumeattachment objects
+	if azVolume.DeletionTimestamp == nil && desiredReplicaCount > currentReplicaCount {
+		klog.Infof("Need %d more replicas for volume (%s)", desiredReplicaCount-currentReplicaCount, azVolume.Spec.UnderlyingVolume)
+		if azVolume.Status.Detail == nil || azVolume.Status.State == v1alpha1.VolumeDeleting || azVolume.Status.State == v1alpha1.VolumeDeleted || azVolume.Status.Detail.ResponseObject == nil {
+			// underlying volume does not exist, so volume attachment cannot be made
+			return nil
+		}
+		if err = r.createReplicas(ctx, min(defaultMaxReplicaUpdateCount, desiredReplicaCount-currentReplicaCount), azVolume.Name, azVolume.Status.Detail.ResponseObject.VolumeID); err != nil {
+			klog.Errorf("failed to create %d replicas for volume (%s): %v", desiredReplicaCount-currentReplicaCount, azVolume.Spec.UnderlyingVolume, err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *ReconcileReplica) getNodesForReplica(ctx context.Context, volumeName string, numReplica int) ([]string, []func(), error) {
+	pods, err := r.controllerSharedState.getPodsFromVolume(volumeName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	unlocks := make([]func(), len(pods))
+	for i, pod := range pods {
+		v, _ := r.controllerSharedState.podLocks.LoadOrStore(pod, &sync.Mutex{})
+		podLock := v.(*sync.Mutex)
+		podLock.Lock()
+		unlocks[i] = podLock.Unlock
+	}
+
+	nodes, err := r.controllerSharedState.getNodesForReplica(ctx, r, nil, pods...)
+	if err != nil {
+		return nil, unlocks, err
+	}
+
+	replicaNodes, err := getNodesWithReplica(ctx, r, volumeName)
+	if err != nil {
+		return nil, unlocks, err
+	}
+
+	skipSet := map[string]bool{}
+	for _, replicaNode := range replicaNodes {
+		skipSet[replicaNode] = true
+	}
+
+	filtered := []string{}
+	numFiltered := 0
+	for _, node := range nodes {
+		if numFiltered >= numReplica {
+			break
+		}
+		if skipSet[node] {
+			continue
+		}
+		filtered = append(filtered, node)
+		numFiltered++
+	}
+
+	return filtered, unlocks, nil
```

Comment:
Is there a reason that we aren't unlocking the pod locks within this function?

Reply:
That is because we don't create replicas within this function. We need the worker to hold onto the lock until it's completed creating the replica CRIs to minimize possibility of replicas of volumes for the same pod being placed in different set of nodes.

Reply:
Makes sense. But having pods are locked in one function but released in another increases the chance of making mistakes. If some other code uses getNodesForReplica and doesn't call unlockPods, the lock on the pods will remain forever.  Maybe we should call getPodsFromVolume and the lock the pods in createReplicas, then call getNodesForReplica from createReplicas passing the pods as a parameter and then release the pod locks in createReplicas.

Reply:
That is a valid concern. I agree that it increases chance of making mistakes. I will go ahead and make necessary changes